pkg/api/steve/catalog/plugin.go

```diff
@@ -0,0 +1,106 @@
+package catalog
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	neturl "net/url"
+	"strings"
+
+	"github.com/gorilla/mux"
+	"github.com/rancher/rancher/pkg/controllers/dashboard/plugin"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+func SetupUIPluginHandlers(router *mux.Router) {
+	router.HandleFunc("/v1/uiplugins", indexHandler)
+	router.HandleFunc("/v1/uiplugins/{name}/{version}/{rest:.*}", pluginHandler)
+}
+
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	authed := isAuthenticated(r)
+	if authed {
+		index, err := json.Marshal(&plugin.Index)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			logrus.Error(err)
+		}
+		w.Write(index)
+	} else {
+		index, err := json.Marshal(&plugin.AnonymousIndex)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			logrus.Error(err)
+		}
+		w.Write(index)
+	}
```

Comment:
Most of the logic is the same, so for simplicity:

```suggestion
	var in *plugin.SafeIndex
	if isAuthenticated(r) {
		in = &plugin.Index
	} else {
		in = &plugin.AnonymousIndex
	}

	index, err := json.Marshal(in)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		logrus.Error(err)
	}
	w.Write(index)
```

Reply:
Done